### PR TITLE
[PKG-2340] imagecodecs 2023.1.23

### PR DESCRIPTION
**Pull request by lorepirri**
[PKG-2340] imagecodecs 2023.1.23

upstream: https://github.com/cgohlke/imagecodecs/blob/v2023.1.23/setup.py
changelog: https://github.com/cgohlke/imagecodecs/blob/v2023.1.23/CHANGES.rst
dependency for: [PKG-2320]`tifffile 2023.4.12`

- bump version to `2023.1.23`
- fix python script
- pin `host` versions
- add `pip` and `pip check` in `test`
- fix `about`
- edit line ending policy in `.gitattributes`
- there is an `ignore_run_exports` section
- patch for `win` on `setup.py` to force pick shared version of `jpeg` for `jpeg8`
- `brotli` was needed strangely just for `win` and `s390x` due to `Cannot open include file: 'brotli/types.h': No such file or directory` see comments. `brunsli` was probably bringing in `brotli` headers for the rest. `brotli` is now enabled for all.
- I had to skip some tests. Most of them are failing due to tolerance not being in the ranges (it might be due to different versions of `numpy`). 
- The following tests fail, not due to tolerance not met:
    - `test_cms_format` for `s390x` and `win` is just failing
    - `test_jpeg2k_numthreads` is disabled for all because it causes seg fault, except for `win`
    - `test_jpeg_rgb_mode` is failing for all.

[PKG-2320]: https://anaconda.atlassian.net/browse/PKG-2320?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PKG-2340]: https://anaconda.atlassian.net/browse/PKG-2340?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml:25]:
```suggestion
```
We do not build cross-compile

File: recipe/meta.yaml
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml:33]:
```suggestion
    - numpy {{ numpy }}
```
Like explained, we have to make sure that the numpy version we use to compile is consistent for a given python version (at least in our ecosystem) because of ABI compatibility.

File: recipe/build.sh
Comment (JeanChristopheMorinPerso) [inline recipe/build.sh:4]:
```suggestion
$PYTHON -m pip install . --no-deps --no-build-isolation -v
```
`-v` will let us see the build logs.

File: recipe/bld.bat
Comment (JeanChristopheMorinPerso) [inline recipe/bld.bat:5]:
```suggestion
%PYTHON% -m pip install . --no-deps --no-build-isolation -v
```

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:12]:
```suggestion
    - 0001-remove-blosc2-codec.patch
```

File: recipe/meta.yaml
Comment (lorepirri) [inline recipe/meta.yaml:48]:
```suggestion
```

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml:2]: I wonder if we could update to the latest version https://github.com/cgohlke/imagecodecs/releases/tag/v2023.7.10. because v2023.1.23 was released 6 months ago

Comment (lorepirri) [reply recipe/meta.yaml:2]: We could, but I noticed that they dropped `py3.8` for that version. Is that OK for us? I wanted to have a recent version with `py3.8` because it's in our range of `py` version. If it's not relevant I will overwrite this and go for the last version or the latest that support `py3.8` which is `2023.3.16` (there should be a PR already there but it's not updated)

Comment (lorepirri) [reply recipe/meta.yaml:2]: I will update this PR to `2023.3.16` with `py3.8`,

once this is merged and done, I will update #13 to [2023.7.10](https://github.com/cgohlke/imagecodecs/releases/tag/v2023.7.10)

Comment (lorepirri) [reply recipe/meta.yaml:2]: `2023.3.16` requires `libjpeg-turbo` which is not in conda `default`. It might be worth having it.

Comment (lorepirri) [reply recipe/meta.yaml:2]: I will not update this here,
will be done in PKG-2501

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml]: ok, thanks

**Comment by JeanChristopheMorinPerso** [comment]
`conda-build` on Windows says:
```
WARNING (imagecodecs): run-exports library package pkgs/main::jpeg-9e-h2bbff1b_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (imagecodecs): run-exports library package pkgs/main::jxrlib-1.1-he774522_2 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```
I think these two warnings should be investigated.

**Comment by lorepirri** [comment]
> `conda-build` on Windows says:
> 
> ```
> WARNING (imagecodecs): run-exports library package pkgs/main::jpeg-9e-h2bbff1b_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
> WARNING (imagecodecs): run-exports library package pkgs/main::jxrlib-1.1-he774522_2 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
> ```
> 
> I think these two warnings should be investigated.

thanks @JeanChristopheMorinPerso I will check them

**Comment by skupr-anaconda** [comment]
> `conda-build` on Windows says:
> 
> ```
> WARNING (imagecodecs): run-exports library package pkgs/main::jpeg-9e-h2bbff1b_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
> WARNING (imagecodecs): run-exports library package pkgs/main::jxrlib-1.1-he774522_2 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
> ```
> 
> I think these two warnings should be investigated.

Regarding jpeg 9e, is that because we build shared and static at the same time https://github.com/AnacondaRecipes/jpeg-feedstock/blob/839f64da35a7f71e9da8349d847254b2b7c7e6ac/recipe/build.sh#L9C21-L10 ?
Screenshot 2023-08-02 at 16 29 32

**Comment by JeanChristopheMorinPerso** [comment]
Alright, I had a talk with @lorepirri and here what should be done to unblock this:

1. Remove `brotli` and `libbrotlicommon` from the `ignore_run_exports` and replace `brotli` by `libbrotlienc` and `libbrotlidec` in the host and run dependencies.
2. On windows, patch https://github.com/cgohlke/imagecodecs/blob/v2023.1.23/setup.py#L148 to use `libjpeg` instead of `jpeg`. `jpeg` is the static library while `libjpeg` is the shared library (as we can see in https://github.com/AnacondaRecipes/jpeg-feedstock/blob/master/recipe/CMakeLists.txt.patch#L23).
3. On windows only, add `jxrlib` to the `ignore_run_exports` section. This is because our Windows variant only provides a static library. Please add a comment.

**Comment by lorepirri** [comment]
> Alright, I had a chat with @lorepirri and here is what should be done to unblock this:
> 
> 1. Remove `brotli` and `libbrotlicommon` from the `ignore_run_exports` and replace `brotli` by `libbrotlienc` and `libbrotlidec` in the host and run dependencies.
> 2. On windows, patch https://github.com/cgohlke/imagecodecs/blob/v2023.1.23/setup.py#L148 to use `libjpeg` instead of `jpeg`. `jpeg` is the static library while `libjpeg` is the shared library (as we can see in https://github.com/AnacondaRecipes/jpeg-feedstock/blob/master/recipe/CMakeLists.txt.patch#L23).
> 3. On windows only, add `jxrlib` to the `ignore_run_exports` section. This is because our Windows variant only provides a static library. Please add a comment.

Thanks a lot for the help @JeanChristopheMorinPerso and the summary.

Windows needs `brotli`, because it fails while looking `Cannot open include file: 'brotli/types.h': No such file or directory`, and adding also `libbrotlicommon` to `libbrotlienc` and `libbrotlidec` did not help).

Tomorrow I will fix the line ending problem with the patch. Locally it works.

As for the `s390x`, I guess it also needs `brotli` as Windows (it seems the only change to me, previously it was building).

File: recipe/meta.yaml
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml:60]: I'm kind of curious as to how it could work on Linux and mac... If the headers are in `brotli`...

Comment (lorepirri) [reply recipe/meta.yaml:62]: @JeanChristopheMorinPerso I wonder the same. Searching through the upstream repo did not help much.

I can see that `brunsli` is set for `[not win and not s390x]`, and both require then `brotli`.

Searching quickly through `brunsli`, I can see that it does bring in `brotli`:
https://github.com/search?q=repo%3Agoogle%2Fbrunsli%20brotli&type=code

**Comment by lorepirri** [comment]
> > `conda-build` on Windows says:
> > ```
> > WARNING (imagecodecs): run-exports library package pkgs/main::jpeg-9e-h2bbff1b_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
> > WARNING (imagecodecs): run-exports library package pkgs/main::jxrlib-1.1-he774522_2 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think these two warnings should be investigated.
> 
> Regarding `jpeg 9e`, is that because we built shared and static libraries at the same time https://github.com/AnacondaRecipes/jpeg-feedstock/blob/839f64da35a7f71e9da8349d847254b2b7c7e6ac/recipe/build.sh#L9C21-L10 ? Screenshot 2023-08-02 at 16 29 32

Thanks @skupr-anaconda ! We were able to fix this. Indeed it was the link to the static version of `jpeg`.

File: recipe/meta.yaml
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:62]: Then we need it as a directly dependency to get the headers. And we'll probably want it as a runtime dep too.

Comment (lorepirri) [reply recipe/meta.yaml:62]: Ok, I will restore `brotli` for all, and restore the runtime dep

**Review by varlackc — APPROVED** [review]
LGTM

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml]: Done, it is building

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Good job @lorepirri!

**Comment by lorepirri** [comment]
> Good job @lorepirri!

Thanks, @JeanChristopheMorinPerso I would not have made it this far without your help and guidance.

Version `2023.3.16` is drafted already in #13, but it requires `libjpeg-turbo 2.1.91 (3.0 beta)` https://github.com/cgohlke/imagecodecs/issues/65#issuecomment-1473856097 and we might want to have `jxrlib` built as shared from now on.  `2023.3.16` is the last version that supports py3.8, see [discussion/comments here](https://github.com/AnacondaRecipes/imagecodecs-feedstock/pull/14/files#r1270451416)

**Comment by JeanChristopheMorinPerso** [comment]
I would reelase 2023.1.23 as is, and then you could work on 2023.3.16 or even 2023.7.10 if we are to go with the latest.

**Comment by lorepirri** [comment]
> I would release 2023.1.23 as is, and then you could work on 2023.3.16 or even 2023.7.10 if we are to go with the latest.

indeed, `2023.1.23` will be out as soon as the second review comes in.

`2023.3.16` is the last version that supports `py3.8` so I would try to build it anyway.

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [inline recipe/meta.yaml:101]: do we want to disable all tests, or would be better to find out which one fails and disable only that one ?

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [inline recipe/meta.yaml:111]: does the "|| true" at make the output always successful ? If so why we need this line as the same on ( without "|| true" ) is on line 105 ?

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:101]: I could investigate further, thanks @M-Waszkiewicz-Anaconda

File: recipe/meta.yaml
Comment (lorepirri) [reply recipe/meta.yaml:111]: The comment above that, on `109`
```
# CAPTURING FAILING S390x tests to be reviewed later (These tests are failing most likely from an upstream Endianess issue)
```
I commented that out because they were run (increasing building time on our CI), they failed anyway, and then the command would give `true`. I think the `conda-forge` feedstock used them for debugging purposes but I would either remove all the lines `109-111` or keep them commented out as a reference.

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml:101]: `[osx and x86_64x]` is enough instead of `[osx and (x86_64 or x86)]` as we already do not support 32-bit architecture https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#preprocessing-selectors

**Review by skupr-anaconda — APPROVED** [review]
LGTM

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Good job @lorepirri and @M-Waszkiewicz-Anaconda ! I know it took a while and a lot of iterations, but I think I'm now happy with the state of this PR.

**Pull request merged**
